kmod/core/core.c

```diff
@@ -341,27 +357,43 @@ static int kpatch_apply_patch(void *data)
 			hash_del_rcu(&func->node);
 		} while_for_each_linked_func();
 
-		return -EBUSY;
+		ret = -EBUSY;
+		goto err;
 	}
 
-	/* run any user-defined load hooks */
+	/* run any user-defined post-patch callbacks */
 	list_for_each_entry(object, &kpmod->objects, list) {
-		if (!kpatch_object_linked(object))
+		if (!kpatch_object_linked(object) ||
+		    !object->post_patch_callback)
 			continue;
-		list_for_each_entry(hook, &object->hooks_load, list)
-			(*hook->hook)();
-	}
 
+		(*object->post_patch_callback)(object);
+	}
 
 	return 0;
+err:
+	kpatch_state_finish(KPATCH_STATE_FAILURE);
+
+	WARN(ret, "error (%d) patching '%s'\n", ret,
+		object->name ? object->name : "vmlinux");
```

Comment:
Unlike livepatch, in kpatch, the `object->name` is `vmlinux`, instead of blank.  So there's no need to check for a NULL name.

Reply:
Also. the WARN() is a bit drastic, and anyway I think it isn't necessary, since the paths branching to here already did pr_err().

Reply:
You're right about `object->name`, I'll make that change.  `WARN()` - I can drop it, I just thought it would be consistent with the module notifier case.

Reply:
But it's different from the module notifier case, because here we can recover from the error by rejecting the patch.